### PR TITLE
PKCS12 certificate support

### PR DESCRIPTION
**Pull request by ALescouet**

### CODE REVIEW

File: openconnect-gp-okta [hunk 1]
```diff
@@ -307,6 +328,10 @@ def main(
         click.echo('--totp-key requires pyotp!', err=True)
         sys.exit(1)
 
+    if cert_path is not None and HAS_PKCS12 is None:
```
Comment (cjriches) [inline openconnect-gp-okta]: `HAS_PKCS12` will never be None, but rather False.

File: openconnect-gp-okta [hunk 2]
```diff
@@ -318,16 +343,51 @@ def main(
...
+    cert_args = []
+    if cert_path is not None:
+        if cert_password_cmd is not None:
+            cert_password = run_cmd(cert_password_cmd, "Cert password")
+        if cert_password is None:
+            cert_password = click.prompt('Cert Password', hide_input=True)
+
+        cert_args = ['-c', cert_path, '-p', cert_password]
+
+        if cert_path.endswith('.p12'):
```
Comment (cjriches) [inline openconnect-gp-okta]: This is the only place `cert_password` is used, so perhaps all the logic to populate that variable should be behind this check?

File: openconnect-gp-okta [hunk 3]
```diff
@@ -318,16 +343,51 @@ def main(
...
+                     tempfile.NamedTemporaryFile(dir=tdir, prefix='key-', buffering=0) as key_file:
+                    crt_file.write(crt)
+                    key_file.write(key)
+                    s.cert = (crt_file.name, key_file.name)
+                    s.mount('https://', TLSAdapter())
+                    saml_req_url = prelogin(s, gateway)
+                    saml_resp_url, saml_resp_data = okta_saml(s, saml_req_url, username, password, factor_priorities, totp_key)
+                    saml_username, prelogin_cookie = complete_saml(s, saml_resp_url, saml_resp_data)
+        else:
+                s.mount('https://', TLSAdapter())
```
Comment (cjriches) [inline openconnect-gp-okta]: nit: weird indentation?

Also, this logic seems to be duplicated. Why do we even need the else branch?

File: openconnect-gp-okta [hunk 4]
```diff
@@ -336,7 +396,7 @@ def main(
         '--user=' + saml_username,
         '--usergroup=gateway:prelogin-cookie',
         '--passwd-on-stdin'
-    ] + list(openconnect_args)
+    ] + list(cert_args) + list(openconnect_args)
```
Comment (cjriches) [inline openconnect-gp-okta]: nit: `cert_args` is already a list

File: openconnect-gp-okta [hunk 5]
```diff
@@ -30,6 +31,17 @@ except ImportError:
     HAS_FIDO2 = False
     UserInteraction = object
 
+HAS_PKCS12 = True
```
Comment (cjriches) [inline openconnect-gp-okta]: nit: it seems slightly neater to me to do

```
try:
    from ... import ...
    HAS_PKCS12 = True
except ImportError:
    HAS_PKCS12 = False

**Comment by cjriches** [comment]
Also, the commit message says

```
-cert-password-cmd: the command that returns the name of the certificate
```

which seems wrong. This is the command that returns the certificate *password*.

File: openconnect-gp-okta (hunk 3 above)
Comment (rickard-faith) [reply openconnect-gp-okta]: In the first case, we wan't the TemporaryDirectory to be cleaned up. The else doesn't need the temporary files.  The common bits could be factored into a function.

**Review by rickard-faith — APPROVED** [review]
In run_cmd, there's a click.echo() that needs to be an f-string.  I couldn't comment on an existing line of code.

File: openconnect-gp-okta (hunk 1 above)
Comment (ALescouet) [reply openconnect-gp-okta]: Done.

File: openconnect-gp-okta (hunk 2 above)
Comment (ALescouet) [reply openconnect-gp-okta]: Done.

File: openconnect-gp-okta (hunk 3 above)
Comment (ALescouet) [reply openconnect-gp-okta]: Done.

File: openconnect-gp-okta (hunk 4 above)
Comment (ALescouet) [reply openconnect-gp-okta]: Done.

File: openconnect-gp-okta (hunk 5 above)
Comment (ALescouet) [reply openconnect-gp-okta]: Done/

**Comment by ALescouet** [comment]
I think I addressed all the comments and will merge it as soon as I get to test it more.

**Review by cjriches — APPROVED** [review]
LGTM, thanks

**Pull request merged**